File: 2058.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        cc = []
        i = 0
        p = head
        while p:
            q = p.next
            if q and q.next:
                if p.val < q.val and q.val > q.next.val or p.val > q.val and q.val < q.next.val:
                    cc.append(i)
            i += 1
            p = q
        
        if len(cc) >= 2:
            return [min(cc[i+1] - cc[i] for i in range(len(cc)-1)), cc[-1] - cc[0]]
        return [-1, -1]

File: 202608/2058.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  vector<int> nodesBetweenCriticalPoints(ListNode *head) {
    int curr = 0;
    vector<int> cps;
    ListNode *p = head;
    while (p && p->next && p->next->next) {
      if (p->val < p->next->val && p->next->val > p->next->next->val ||
          p->val > p->next->val && p->next->val < p->next->next->val) {
        cps.push_back(curr);
      }
      ++curr;
      p = p->next;
    }
    vector<int> ans{-1, -1};
    if (cps.size() < 2)
      return ans;
    ans[1] = cps[cps.size() - 1] - cps[0];
    ans[0] = INT_MAX;
    for (int i = 1; i < cps.size(); ++i) {
      ans[0] = min(ans[0], cps[i] - cps[i - 1]);
    }
    return ans;
  }
};
